src/codegen.cpp

```diff
@@ -5169,7 +5169,23 @@ static Function* gen_cfun_wrapper(
         // add nest parameter (pointer to jl_value_t* data array) after sret arg
         assert(closure_types);
         std::vector<Type*> fargt_sig(sig.fargt_sig);
+
         fargt_sig.insert(fargt_sig.begin() + sig.sret, T_pprjlvalue);
+        // Shift LLVM attributes for parameters one to the right, as
+        // we are adding the extra nest parameter after sret arg.
+        // AttributeList has function attributes and return value
+        // attributes before the parameter attributes.
+        if (attributes.getNumAttrSets() > static_cast<unsigned>(2 + sig.sret)) {
+            AttrBuilder toShift;
+            // Skip past function and return and sret attributes to the first real parameter
+            for (auto it = attributes.index_begin() + 2 + sig.sret; it != attributes.index_end(); ++it) {
+                AttrBuilder toShiftTemp(attributes.getAttributes(it));
+                attributes = attributes.setAttributes(jl_LLVMContext, it, toShift);
+                toShift = std::move(toShiftTemp);
+            }
+            attributes = attributes.setAttributes(jl_LLVMContext, attributes.index_end(), toShift);
```

Comment:
oh, oops, this has the wrong type now
```suggestion
            AttributeSet toShift;
            // Skip past function and return and sret attributes to the first real parameter
            for (auto it = attributes.index_begin() + 2 + sig.sret; it != attributes.index_end(); ++it) {
                AttributeSet toShiftTemp(attributes.getAttributes(it));
                attributes = attributes.setAttributes(jl_LLVMContext, it, toShift);
                toShift = toShiftTemp;
            }
            attributes = attributes.setAttributes(jl_LLVMContext, attributes.index_end(), toShift);
```

Reply:
Whoops, and I noticed now, we of course also need to remove the attributes form the first parameter. I made another version which I think is efficient enough.

Reply:
I think that this did–the first was set to `toShift = AttributeSet()`